[13.0] Create module product_total_weight_from_packaging

grindtildeath opened the pull request:
Depends on:
 - [x] https://github.com/OCA/stock-logistics-warehouse/pull/939
 - [x] https://github.com/Tonow-c2c/stock-logistics-warehouse/pull/1


This module provides a weight calculation function for products based on their
packagings weight.

It uses module `stock_packaging_calculator` to get weight from packagings
having a weight defined first and fallback on product weight field if no
weight is defined on the packagings.

grindtildeath committed:
Create module product_total_weight_from_packaging

simahawk commented on product_total_weight_from_packaging/models/product.py:
```diff
@@ -0,0 +1,35 @@
+        total_weight = 0
+        for packaging_dict in qty_by_packaging_with_weight:
+            if packaging_dict.get("is_unit"):
+                pack_weight = self.weight
```
is this needed? You set the weight for units on line 30 as `self.weight` as well. :thinking:

simahawk commented:
the 2 pending PRs got merged

jbaudoux commented:
Just a small remark here : https://github.com/OCA/stock-logistics-warehouse/pull/939#issuecomment-660894458

jbaudoux commented on product_total_weight_from_packaging/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+{
+    "name": "Product Total Weight From Packaging",
+    "summary": "Compute product quantity to pick by packaging",
```
Wrong summary

Compute estimated weight based on product's packaging weights

jbaudoux commented on product_total_weight_from_packaging/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,6 @@
+This module provides a weight calculation function for products based on their
```
```suggestion
This module provides a function to estimate the weight of a given quantity taking into account the product's```

jbaudoux reviewed (commented):
Small description remarks

simahawk commented:
@grindtildeath can you squash and add the dependency on stock-logistics-wh?

rousseldenis commented on product_total_weight_from_packaging/models/product.py:
```diff
@@ -0,0 +1,34 @@
+
+    def _prepare_qty_by_packaging_values(self, packaging_tuple, qty_per_pkg):
+        res = super()._prepare_qty_by_packaging_values(packaging_tuple, qty_per_pkg)
+        if self.env.context.get("_with_packaging_weight"):
```
@grindtildeath Why doing some complex things with context here as you add just an entry in the result? There is no behaviour bypassing of the original packaging caclulator.

grindtildeath replied:
As the module can be used in different contexts, I thought we can avoid having the weight in the result if unneeded, but you're right in the sense having this value in the dict doesn't break anything from original function.
@simahawk what d'you think?

simahawk replied:
@grindtildeath thinking about this twice in the ctx where is needed: do we really need the weight to be in `product_qty_by_packaging` result?
Because your goal here can be achieved with `sum([x.qty * x.max_weight for x in self.packaging_ids.filtered(lambda p: p.qty and p.max_weight))])` no?

Regarding adding `weight` to the original result w/out conditions... fine if we do it in the base module. You'll have to add a dependency on delivery.
And yes, you can break things - like tests - if you add it always and it's not foreseen.

grindtildeath replied:
@simahawk Did you check the test and usage before suggesting this? :roll_eyes:

simahawk replied:
you have one test that only checks the result of `get_total_weight_from_packaging`.
How should I imagine something different?
In my suggestion you only miss the unit weight because I omitted it. You can add it very easily, no?
What else do I miss?

grindtildeath replied:
@simahawk The test is as follows:

 - we have a product with a weight of 5 and following packagings
    - Pair (qty=2, max_weight=12.5)
    - Cardbox (qty=10, max_weight=55)
    - Pallet (qty=200, max_weight=0)

As stated in the module readme:
>This module provides a a function to estimate the weight
of a given quantity of a product, taking into account the product
packaging's weights.

> It uses module `stock_packaging_calculator` to get weight from product packagings
having a weight defined first and fallback on product weight field if no
weight is defined on any of the packagings.

Therefore, if we want to compute the weight for 259 units, we must do 25 * 55 + 4 * 12.5 + 5, because the pallet doesn't have a weight defined.

Anyway, I'm sorry but what does `sum([x.qty * x.max_weight for x in self.packaging_ids.filtered(lambda p: p.qty and p.max_weight))])` actually do? 
I'm still trying to figure out what you can achieve by multiplicating the quantity inside a packaging by the weight of said packaging...

If I have a 6 pack of beers and each 0.5l beer weights 0.5kg, my 6-pack should have a max weight around 3kg. And I really don't understand how your suggestion of multiplicating 3kg * 6pieces would give anything meaningful...

simahawk replied:
Sorry, don't take my example literally. I meant "loop on them, filter them, use them as you want" but I admit that was not crystal clear :sweat_smile: 

I'm just saying that maybe you don't need that module as a base and you can rely on your own computation and/or use some bits only from that module (eg: _qty_by_pkg) and do what you want here.

Another options is, instead of overriding `_prepare_qty_by_packaging_values` and check for ctx key, you could improve `_product_qty_by_packaging` to lookup for `_packaging_values_handler` and pass your own handler.
This way the base module is even more pluggable.

Eg:

`value = self._prepare_qty_by_packaging_values(pkg, qty_per_pkg)`
becomes
```py
prepare_values = self.env.context.get('_packaging_values_handler', self._prepare_qty_by_packaging_values)
value = prepare_values(pkg, qty_per_pkg)
```
Then you use your own handler here. No overlap of features/data/whatever, no useless checks.

That being said. I don't care that much here about cosmetics and alike. That's why I approved it before.
The module is alpha: if it works and does not break anything, it's ok. Maybe just add a TODO comment there to improve it on the next round.

simahawk replied:
> > can you squash and add the dependency on stock-logistics-wh?
> 
> @simahawk shouldn't we move the module here to avoid this dependency? wink :trollface:

go ahead if you have time. But if you change the module's name, add a migration script ;)

simahawk replied:
as I was touching the module, I added the custom value handler hook here https://github.com/OCA/stock-logistics-warehouse/pull/949

grindtildeath commented:
> can you squash and add the dependency on stock-logistics-wh?

@simahawk shouldn't we move the module here to avoid this dependency? :wink:  :trollface:

simahawk commented:
> @simahawk shouldn't we move the module here to avoid this dependency? wink :trollface:

go ahead if you have time. But if you change the module's name, add a migration script ;)

rousseldenis commented:
@grindtildeath Will you fix this ?

jgrandguillaume reviewed (approved):
LGTM, I suggest to merge this work as it is Alpha we can improve the rest on a later stage.

jgrandguillaume commented:
/ocabot merge nobump

jgrandguillaume commented:
Please add the oca_dependencies @grindtildeath

grindtildeath committed:
Use new values handler

grindtildeath committed:
Add oca_dependencies.txt

grindtildeath committed:
packaging_uom: Remove readonly on packaging.qty

Removing the readonly flag on the qty field avoids to trigger the
inverse method and loose the value of qty when it's passed to create.

grindtildeath commented:
@rousseldenis Can you please check last commit?

guewen commented:
/ocabot merge patch
